### DISCUSSION TITLE
Allow creating open delegates from instance method groups

### DISCUSSION THREAD

**Discussion by tmat** [dotnet/csharplang#4303] [Language Ideas]
Currently C# does not allow to create an open delegate given an method group that has a matching instance method overload like so: 

```C#
new Func<string, string, bool>(string.Contains)
```

```
error CS0123: No overload for 'Contains' matches delegate 'Func<string, string, bool>'
```

The Reflection API allows such delegate to be created using:

```C#
typeof(string).GetMethod("Contains").CreateDelegate(typeof(Func<string, string, bool>))
```

The following should also work:

```C#
new Func<string, string>(string.ToLowerInvariant)
```

The reflection equivalent works:

```C#
typeof(string).GetMethod("ToLowerInvariant").CreateDelegate(typeof(Func<string, string>));
```

**Comment by huoyaoyuan** [comment]
The valid syntax with same semantics should be `(x, y) => x.Contains(y)`. It will require an inlining, the open delegate version will require an IL stub to move the ABI, so I'd expect their codegen to be idential.

**Comment by tmat** [comment]
I'd also like this to work:

```C#
new Func<string, string>(string.ToLowerInvariant)
```

The reflection equivalent works:

```C#
typeof(string).GetMethod("ToLowerInvariant").CreateDelegate(typeof(Func<string, string>));
```

[Updated the original text.]

**Comment by tmat** [comment]
@huoyaoyuan Yes, you can write lambda but that's extra clutter. You'd probably want to use `static` to ensure its closure-free: `static (x, y) => x.Contains(y)`. It's better to just type `string.Contains`.
